source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -4351,6 +4351,148 @@ export class ColumnSet extends CardElementContainer {
+            if (s.mimeType) {
+                let mimeComponents = s.mimeType.split('/');
+                if (mimeComponents.length == 2) {
+                    return typeFilter(mimeComponents[0]);
+                }
+            }
+            return false;
+        });
+    }
+
+    private makeButton(label: string): HTMLAnchorElement {
+        let button = document.createElement("a");
+        button.className = "ac-media-button";
+        button.text = label;
+        return button;
+    }
+
+    protected internalRender(): HTMLElement {
+        var element = document.createElement("div");
+
+        element.className = this.hostConfig.makeCssClassName("ac-media");
+
+        let goodSources = this.processSources();
+
+        if (goodSources.length) {
+
+            this._mediaElement = <HTMLMediaElement>document.createElement(this._mediaType);
+            this._mediaElement.style.width = "100%";
+
+            goodSources.forEach(s => {
+                let src: HTMLSourceElement = document.createElement("source");
+                src.src = s.url;
+                src.type = s.mimeType;
+                this._mediaElement.appendChild(src);
+            });
+            element.appendChild(this._mediaElement);
+
+            var play_button = this.makeButton("▶️");
+            play_button.onclick = e => this.play();
+            element.appendChild(play_button);
+
+            var pause_button = this.makeButton("⏸️");
+            pause_button.onclick = e => this.pause();
+            element.appendChild(pause_button);
+
+            var stop_button = this.makeButton("⏹️");
+            stop_button.onclick = e => this.stop();
+            element.appendChild(stop_button);
+        }
+
+        return element;
+    }
+
+    play() {
+        this._mediaElement.play();
+    }
+
+    stop() {
+        this._mediaElement.load();
+        this._mediaElement.pause();
+    }
+
+    pause() {
+        this._mediaElement.pause();
+    }
+
+    toJSON() {
+        let result = super.toJSON();
+
+        Utils.setProperty(result, "poster", this.poster);
+        Utils.setProperty(result, "altText", this.altText);
+        if (this.sources.length) {
+            var sources = [];
+
+            for (let source of this.sources) {
+                sources.push(source.toJSON());
+            }
+
+            Utils.setProperty(result, "sources", sources);
+        }
+
+        return result;
+    }
+
+    parse(json: any) {
+        super.parse(json);
+
+        this.poster = json.poster;
+        this.altText = json.altText;
+
+        if (json["sources"] != undefined) {
+            var sourceArray = json["sources"] as Array<any>;
+            this.sources = sourceArray.map((json: any) => new Source().parse(json));
+        }
+    }
+
+    getJsonTypeName(): string {
+        return "Media";
+    }
+
+    renderSpeech(): string {
+        if (this.speak != null)
+            return this.speak + '\n';
```

Comment:
Please always use open and close curly brackets, even if there is only one statement.

Reply:
I won't fall for the temptation of saying "I copied from elsewhere on the same file" because it's not a justification.

Reply:
Feel free to fix that elsewhere as well ;)